rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/Rename.rsc

```diff
@@ -0,0 +1,400 @@
+    }
+
+    return names;
+}
+
+Maybe[loc] locationOfName(Name n) = just(n.src);
+Maybe[loc] locationOfName(QualifiedName qn) = just((qn.names[-1]).src);
+Maybe[loc] locationOfName(FunctionDeclaration f) = just(f.signature.name.src);
+Maybe[loc] locationOfName(Variable v) = just(v.name.src);
+Maybe[loc] locationOfName(Declaration d) = just(d.name.src) when d is annotation
+                                                              || d is \tag;
+Maybe[loc] locationOfName(Declaration d) = locationOfName(d.user.name) when d is \alias
+                                                                         || d is dataAbstract
+                                                                         || d is \data;
+Maybe[loc] locationOfName(TypeVar tv) = just(tv.name.src);
+Maybe[loc] locationOfName(Header h) = locationOfName(h.name);
+default Maybe[loc] locationOfName(Tree t) = nothing();
+
+private tuple[set[IllegalRenameReason] reasons, list[TextEdit] edits] computeTextEdits(WorkspaceInfo ws, start[Module] m, set[loc] defs, set[loc] uses, str name) {
+    if (reasons := collectIllegalRenames(ws, m, defs, uses, name), reasons != {}) {
+        return <reasons, []>;
+    }
+
+    replaceName = escapeName(name);
+    return <{}, [replace(l, replaceName) | l <- findNames(m, defs + uses)]>;
+}
+
+private tuple[set[IllegalRenameReason] reasons, list[TextEdit] edits] computeTextEdits(WorkspaceInfo ws, loc moduleLoc, set[loc] defs, set[loc] uses, str name) =
+    computeTextEdits(ws, parseModuleWithSpaces(moduleLoc), defs, uses, name);
+
+private bool rascalMayOverloadSameName(set[loc] defs, map[loc, Define] definitions) {
+    set[str] names = {definitions[l].id | l <- defs, definitions[l]?};
+    if (size(names) > 1) return false;
+
+    map[loc, Define] potentialOverloadDefinitions = (l: d | l <- definitions, d := definitions[l], d.id in names);
+    return rascalMayOverload(defs, potentialOverloadDefinitions);
+}
+
+private bool isFunctionLocalDefs(WorkspaceInfo ws, set[loc] defs) {
+    for (d <- defs) {
+        if (Define _: <_, _, _, _, funDef, defType(afunc(_, _, _))> <- ws.defines
+          , isContainedIn(ws.definitions[d].scope, funDef)) {
+            continue;
+        }
+        return false;
+    }
+    return true;
+}
+
+private bool isFunctionLocal(WorkspaceInfo ws, cursor(def(), cursorLoc, _)) =
+    isFunctionLocalDefs(ws, getOverloadedDefs(ws, {cursorLoc}, rascalMayOverloadSameName));
+private bool isFunctionLocal(WorkspaceInfo ws, cursor(use(), cursorLoc, _)) =
+    isFunctionLocalDefs(ws, getOverloadedDefs(ws, getDefs(ws, cursorLoc), rascalMayOverloadSameName));
+private bool isFunctionLocal(WorkspaceInfo _, cursor(typeParam(), _, _)) = true;
+private bool isFunctionLocal(WorkspaceInfo _, cursor(collectionField(), _, _)) = false;
+private bool isFunctionLocal(WorkspaceInfo _, cursor(moduleName(), _, _)) = false;
```

Comment:
A `default` is missing. The patterns cover all possibilities now, but could the `CursorKind` be extended in the future?

Reply:
Added a default